File: src/main/java/com/atsuishio/superbwarfare/compat/tacz/TACZGunEventHandler.java
package com.atsuishio.superbwarfare.compat.tacz;

import net.fabricmc.loader.api.FabricLoader;

public class TACZGunEventHandler {

    /*
    public static void entityHurtByTACZGun(EntityHurtByGunEvent.Pre event) {
        if (event.getHurtEntity() instanceof VehicleEntity) {
            event.setHeadshot(false);
        }
    }

    public static boolean hasMod() {
        return FabricLoader.getInstance().isModLoaded("tacz");
    }

    public static boolean compatCondition() {
        var modFile = LoadingModList.get().getModFileById("tacz");
        if (modFile == null) return false;
        DefaultArtifactVersion modVersion = new DefaultArtifactVersion(modFile.versionString());
        return modVersion.compareTo(new DefaultArtifactVersion("1.1.4")) >= 0;
    }

    public static ResourceLocation getTaczCompatIcon(ItemStack stack) {
        if (stack.getItem() instanceof IGun iGun) {
            ResourceLocation gunId = iGun.getGunId(stack);
            GunData gunData = TimelessAPI.getClientGunIndex(gunId).map(ClientGunIndex::getGunData).orElse(null);
            GunDisplayInstance display = TimelessAPI.getGunDisplay(stack).orElse(null);
            if (gunData != null && display != null) {
                return display.getHUDTexture();
            }
        }
        return null;
    }
    */
}

File: src/main/kotlin/com/atsuishio/superbwarfare/compat/tacz/TACZGunEventHandler.kt
package com.atsuishio.superbwarfare.compat.tacz

import com.atsuishio.superbwarfare.entity.vehicle.base.VehicleEntity
import com.tacz.guns.api.TimelessAPI
import com.tacz.guns.api.event.common.EntityHurtByGunEvent
import com.tacz.guns.api.item.IGun
import net.minecraft.resources.ResourceLocation
import net.minecraft.world.item.ItemStack

object TACZGunEventHandler {
    fun registerEvents() {
        EntityHurtByGunEvent.PRE.register(::entityHurtByTACZGun)
    }

    fun entityHurtByTACZGun(event: EntityHurtByGunEvent.Pre) {
        if (event.hurtEntity is VehicleEntity) {
            event.setHeadshot(false)
        }
    }

    fun hasMod(): Boolean {
        return TaczCompat.isLoaded()
    }

    fun compatCondition(): Boolean {
        return TaczCompat.isCompatible()
    }

    fun getTaczCompatIcon(stack: ItemStack): ResourceLocation? {
        val item = stack.item
        if (item is IGun) {
            val gunId: ResourceLocation = item.getGunId(stack)
            val gunData = TimelessAPI.getClientGunIndex(gunId)
                .map { obj -> obj.gunData }.orElse(null)
            val display = TimelessAPI.getGunDisplay(stack).orElse(null)
            if (gunData != null && display != null) {
                return display.hudTexture
            }
        }
        return null
    }
}
